tools/xdp/s2n-quic-xdp/src/ring/cursor.rs

```diff
@@ -0,0 +1,488 @@
+        if cfg!(debug_assertions) {
+            let max_len = self.cached_consumer_len();
+            assert!(
+                max_len >= len,
+                "available: {}, requested: {}, {self:?}",
+                max_len,
+                len
+            );
+        }
+        self.cached_consumer += len;
+        self.cached_len -= len;
+        self.consumer().fetch_add(len, Ordering::Release);
+    }
+
+    /// Returns `true` if the ring needs to be notified when entries are updated
+    ///
+    /// See [xsk.h](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/headers/xdp/xsk.h#L87).
+    #[inline]
+    pub fn needs_wakeup(&self) -> bool {
+        self.flags().contains(RingFlags::NEED_WAKEUP)
+    }
+
+    /// Returns a reference to the flags on the ring
+    #[inline]
+    pub fn flags(&self) -> &RingFlags {
+        unsafe { &*self.flags.as_ptr() }
+    }
+
+    /// Returns the current consumer entries
+    ///
+    /// # Safety
+    ///
+    /// This function MUST only be used by the consumer side.
+    #[inline]
+    pub unsafe fn consumer_data(&mut self) -> (&mut [T], &mut [T]) {
+        let idx = self.cached_consumer();
+        let len = self.cached_len;
+
+        debug_assert_eq!(len, self.cached_consumer_len());
+
+        self.mut_slices(idx as _, len as _)
+    }
+
+    /// Returns the current producer entries
+    ///
+    /// # Safety
+    ///
+    /// This function MUST only be used by the producer side.
+    #[inline]
+    pub unsafe fn producer_data(&mut self) -> (&mut [T], &mut [T]) {
+        let idx = self.cached_producer();
+        let len = self.cached_len;
+
+        debug_assert_eq!(len, self.cached_producer_len());
+
+        self.mut_slices(idx as _, len as _)
+    }
+
+    /// Creates a pair of slices for a given cursor index and len
+    #[inline]
+    fn mut_slices(&mut self, idx: u64, len: u64) -> (&mut [T], &mut [T]) {
+        if len == 0 {
+            return (&mut [][..], &mut [][..]);
+        }
+
+        let ptr = self.desc.as_ptr() as *mut T;
+
+        if let Some(tail_len) = (idx + len).checked_sub(self.size as _) {
+            let head_len = self.size as u64 - idx;
+            debug_assert_eq!(head_len + tail_len, len);
+            let head = unsafe { core::slice::from_raw_parts_mut(ptr.add(idx as _), head_len as _) };
+            let tail = unsafe { core::slice::from_raw_parts_mut(ptr, tail_len as _) };
+            (head, tail)
+        } else {
+            let slice = unsafe { core::slice::from_raw_parts_mut(ptr.add(idx as _), len as _) };
+            (slice, &mut [][..])
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bolero::{check, generator::*};
+    use core::cell::UnsafeCell;
+
+    #[derive(Clone, Copy, Debug, TypeGenerator)]
+    enum Op {
+        ConsumerAcquire(u16),
+        ConsumerRelease(u16),
+        ProducerAcquire(u16),
+        ProducerRelease(u16),
```

Comment:
would testing with u32 be useful since some of the values on Cursor are u32

Reply:
Not in this case since we're limiting the queue size for the sake of memory usage. Having the smaller lengths also consumes less of the fuzzer input so the corpus is smaller as well.

https://github.com/aws/s2n-quic/blob/7e3c5b520757e21deb800a3dccc302bfc618c954/tools/xdp/s2n-quic-xdp/src/ring/cursor.rs#L485